// @ndn/keychain/lib/store/stores_browser.js
import { createStore, del, get, keys, set } from "idb-keyval";
import { CertStore } from "./cert-store_browser.js";
import { KeyStore } from "./key-store_browser.js";
class IdbStoreProvider {
    // Firefox does not support structured clone of ECDSA CryptoKey.
    // https://bugzilla.mozilla.org/show_bug.cgi?id=1545813
    canSClone = !/rv:.*Gecko\//.test(navigator.userAgent);
    store;
    constructor(dbName) {
        this.store = createStore(dbName, "");
    }
    list() {
        return keys(this.store);
    }
    async get(key) {
        const value = await get(key, this.store);
        if (value === undefined) {
            throw new Error(`${key} does not exist`);
        }
        return value;
    }
    insert(key, value) {
        return set(key, value, this.store);
    }
    erase(key) {
        return del(key, this.store);
    }
}
export function openStores(locator, algoList) {
    return [
        new KeyStore(new IdbStoreProvider(`${locator} e3617e69-4f2c-4221-955a-bea86832595f`), algoList),
        new CertStore(new IdbStoreProvider(`${locator} 9503f5a4-a0a3-4cb0-b764-b0e78afd4ada`)),
    ];
}
