// @strattadb/environment/dist/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
Object.defineProperty(exports, "Parser", {
  enumerable: true,
  get: function () {
    return _parsers.Parser;
  }
});
Object.defineProperty(exports, "makeEnv", {
  enumerable: true,
  get: function () {
    return _environment.makeEnv;
  }
});
Object.defineProperty(exports, "Env", {
  enumerable: true,
  get: function () {
    return _environment.Env;
  }
});
Object.defineProperty(exports, "Schema", {
  enumerable: true,
  get: function () {
    return _environment.Schema;
  }
});
Object.defineProperty(exports, "SchemaEntry", {
  enumerable: true,
  get: function () {
    return _environment.SchemaEntry;
  }
});
Object.defineProperty(exports, "SchemaEntryRequiredInfo", {
  enumerable: true,
  get: function () {
    return _environment.SchemaEntryRequiredInfo;
  }
});
Object.defineProperty(exports, "EnvironmentVariableError", {
  enumerable: true,
  get: function () {
    return _EnvironmentVariableError.default;
  }
});
exports.parsers = void 0;

var _parsers = require("./parsers");

var _environment = require("./environment");

var _EnvironmentVariableError = _interopRequireDefault(require("./EnvironmentVariableError"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const parsers = {
  array: _parsers.array,
  boolean: _parsers.boolean,
  email: _parsers.email,
  float: _parsers.float,
  integer: _parsers.integer,
  url: _parsers.url,
  ipAddress: _parsers.ipAddress,
  negativeInteger: _parsers.negativeInteger,
  nonNegativeInteger: _parsers.nonNegativeInteger,
  nonPositiveInteger: _parsers.nonPositiveInteger,
  port: _parsers.port,
  positiveInteger: _parsers.positiveInteger,
  regex: _parsers.regex,
  string: _parsers.string,
  whitelist: _parsers.whitelist
};
exports.parsers = parsers;
