// @ndn/dpdkmgmt/lib/prefix-reg_browser.js
import { ReadvertiseDestination } from "@ndn/fw";
import { gql } from "graphql-request";
/** Enable prefix registration via NDN-DPDK GraphQL management API. */
export class NdndpdkPrefixReg extends ReadvertiseDestination {
    client;
    faceID;
    constructor(client, faceID) {
        super();
        this.client = client;
        this.faceID = faceID;
    }
    async doAdvertise(name, state) {
        const resp = await this.client.request(gql `
      mutation insertFibEntry($name: Name!, $nexthops: [ID!]!, $strategy: ID) {
        insertFibEntry(name: $name, nexthops: $nexthops, strategy: $strategy) {
          id
        }
      }
    `, {
            name: name.toString(),
            nexthops: [this.faceID],
        });
        state.fibEntryID = resp.insertFibEntry.id;
    }
    async doWithdraw(name, state) {
        void name;
        if (!state.fibEntryID) {
            return;
        }
        await this.client.request(gql `
      mutation delete($id: ID!) {
        delete(id: $id)
      }
    `, {
            id: state.fibEntryID,
        });
        delete state.fibEntryID;
    }
}
